### PR TITLE
Show results immediately if user hits "back" key after finishing gameplay

### LINKED ISSUE #18089: Player should not be able to exit gameplay after progression to results has begun
Currently, if you hit the exit key (after setting hold delay to zero) directly after passing a beatmap, the player will not reach the results screen. This feels like pretty bad UX.

Optimally, during this previous, invoking the pause menu should still be feasible. I did look into doing this but it's a bit involved due to the `Task` which is used to progress to the next screen.

Failing that, a better-than-nothing option would be to have it immediately switch to the results screen (similar to stable), or outright block the exit sequence during this period (least preferable).

### Discussed in https://github.com/ppy/osu/discussions/17955

<div type='discussions-op-text'>

<sup>Originally posted by **respektive** April 24, 2022</sup>
As has been reported before: https://github.com/ppy/osu/issues/3060 https://github.com/ppy/osu/issues/13644

I believe the behaviour from stable should come back for this as I see a lot of players do this all the time and I don't think we can expect people to break this habit that easily. Even after playing more and more lazer over the last months, it is very hard for me to get rid of this habit and I still find myself hitting the escape key after almost every play. 

Maybe this isn't affecting as many players as I think though, as it's not that much of a problem with the "hold to confirm" option, which i personally have disabled. I'm interested to hear more opinions on this.</div>

### PR DESCRIPTION
**Pull request by peppy**
Closes https://github.com/ppy/osu/issues/18089.

I've gone ahead and matched the osu!stable behaviour for this, as it seems like it's what people are used to and they will settle for no less.

### CODE REVIEW

**Comment by bdach** [comment]
Two things:

First one, minor. With this change, do we still need the following code? It's not being removed in this pull.

https://github.com/ppy/osu/blob/d2044fc2ccb3df48a103865bbe44b577553a3c8c/osu.Game/Screens/Play/Player.cs#L570-L580

Second, probably more important. If a storyboard has an outro, exiting during outro will not progress to results due to:

https://github.com/ppy/osu/blob/d2044fc2ccb3df48a103865bbe44b577553a3c8c/osu.Game/Screens/Play/Player.cs#L737-L743

I haven't checked what stable does but the fact that skipping outro will get you to results but exiting kicks to song select feels quite bad regardless.

**Comment by peppy** [comment]
> With this change, do we still need the following code?

Likely not. I noticed this but forgot to check whether it can be removed.

> If a storyboard has an outro, exiting during outro will not progress to results due to..

Should definitely use the new behaviour instead.

File: osu.Game/Screens/Play/Player.cs [hunk 1]
```diff
@@ -660,10 +671,10 @@ protected void SetGameplayStartTime(double time)
         /// <remarks>This can be called from a child screen in order to trigger the restart process.</remarks>
         /// </summary>
         /// <param name="quickRestart">Whether a quick restart was requested (skipping intro etc.).</param>
-        public void Restart(bool quickRestart = false)
+        public bool Restart(bool quickRestart = false)
```
Comment (akiya-nagatsuka) [inline osu.Game/Screens/Play/Player.cs:674]: Sorry for a nitpick: 
also need to add `/// <returns>` doc like for `PerformExit`?

**Comment by peppy** [comment]
> I can't actually hit this path but it seems like it could be removable (just do nothing if screen isn't current).

It can be removed, but requires some (probably fine) changes to the restart process, which may add a bit of mental burden to the review process (see https://github.com/ppy/osu/pull/25097/commits/d174a6ce613c39efd24ec8167288a453d58dd90f).

> Second, probably more important. If a storyboard has an outro, exiting during outro will not progress to results

Have fixed this, a very easy fix luckily. It even had test coverage.

**Review by bdach — APPROVED** [review]
I _think_ this is okay, but leaning heavily on tests passing here...

**Pull request merged**